### ISSUE TITLE
Initial value delayed in waitRace operator

### ISSUE THREAD

**Issue by vitaly-t** [vitaly-t/iter-ops#182]
Current implementation (v2.2.1) of the new operator [waitRace] delays initial resolution when the source is slower than the resolution speed.

The example below demonstrates that issue...

```ts
import {delay, map, pipeAsync, waitRace} from 'iter-ops';

function* gen(n) {
    while (n > 0) {
        yield n--;
    }
}

const i = pipeAsync(gen(10), delay(1000), map(a => Promise.resolve(a)), waitRace(5));

(async function () {
    for await(const a of i) {
        console.log(a);
    }
})();
```

In the example, the first value is expected to be resolved after 1s, but instead it is resolved after 2s, i.e. waiting for the next value to be resolved first, which is wrong. Something is wrong in the implementation logic.

For further details, see [this discussion](https://github.com/vitaly-t/iter-ops/discussions/179).

[waitRace]:https://vitaly-t.github.io/iter-ops/functions/waitRace.html

---

@RebeccaStevens this is a tricky one, could really use some help here ;)

@bergus I have tried to apply [your solution from StackOverflow](https://stackoverflow.com/questions/74293770/unordered-resolution-of-a-list-of-promises), but wasn't successful. Maybe it can be reduced to a simpler fix on the existing code?

**Comment by vitaly-t** [comment]
[I have added a test](https://github.com/vitaly-t/iter-ops/blob/main/test/ops/wait-race/async.ts#L38) that can be uncommented once the fix is in place:

```ts
    it('must start resolving without delay', async () => {
        const input = [1, 2, 3, 4, 5, 6, 7];
        const output: {value: number; delay: number}[] = [];
        const i = pipeAsync(
            input,
            delay(50),
            map((a) => Promise.resolve(a)),
            waitRace(5)
        );
        const start = Date.now();
        for await (const value of i) {
            const delay = Date.now() - start;
            output.push({value, delay});
        }
        /*
        // TODO: This test must pass once issue #182 has been resolved
        expect(output[0].delay).to.be.lessThan(75); // must be between 100 and 200
        expect(output[1].delay).to.be.lessThan(125); // must be between 200 and 300
         */
    });
```

**Comment by vitaly-t** [comment]
I have improved [the test](https://github.com/vitaly-t/iter-ops/blob/main/test/ops/wait-race/async.ts#L38) further, to make it more comprehensive.

**Comment by vitaly-t** [comment]
I made a couple more attempts at fixing it, without success again. It would take a special inspiration to implement it properly, it seems.

@RebeccaStevens Could really use some help here, as I'm stuck with this issue.

**Comment by RebeccaStevens** [comment]
Looking into it. I've kind of got it working but not really.

**Comment by vitaly-t** [comment]
Well, if it can pass [the test](https://github.com/vitaly-t/iter-ops/blob/main/test/ops/wait-race/async.ts#L38), that's a good sign ;)

**Comment by bergus** [comment]
> @bergus I have tried to apply [your solution from StackOverflow](https://stackoverflow.com/questions/74293770/unordered-resolution-of-a-list-of-promises), but wasn't successful.

Aw. It seems to work fine for me:
```js
class AsyncQueue {
    constructor() {
        // invariant: at least one of the arrays is empty.
        // when `resolvers` is `null`, the queue has ended.
        this.resolvers = [];
        this.promises = [];
    }
    putNext(result) {
        if (!this.resolvers)
            throw new Error('Queue already ended');
        if (this.resolvers.length)
            this.resolvers.shift()(result);
        else
            this.promises.push(Promise.resolve(result));
    }
    put(value) {
        this.putNext({ done: false, value });
    }
    end() {
        for (const res of this.resolvers)
            res({ done: true, value: undefined });
        this.resolvers = null;
    }
    next() {
        if (this.promises.length)
            return this.promises.shift();
        else if (this.resolvers)
            return new Promise(resolve => { this.resolvers.push(resolve); });
        else
            return Promise.resolve({ done: true, value: undefined });
    }
    [Symbol.asyncIterator]() {
        // Todo: Use AsyncIterator.from()
        return this;
    }
}
function limitConcurrent(iterable, n) {
    const produced = new AsyncQueue();
    const consumed = new AsyncQueue();
    (async () => {
        try {
            let count = 0;
            for await (const p of iterable) {
                const promise = Promise.resolve(p.nested);
                promise.then(value => {
                    produced.put(value);
                }, _err => {
                    produced.putNext(promise); // with rejection already marked as handled
                });
                if (++count >= n) {
                    await consumed.next(); // happens after any produced.put[Next]()
                    count--;
                }
            }
            while (count) {
                await consumed.next(); // happens after any produced.put[Next]()
                count--;
            }
        }
        catch (e) {
            // ignore `iterable` errors?
        }
        finally {
            produced.end();
        }
    })();
    return (async function* () {
        for await (const value of produced) {
            yield value;
            consumed.put();
        }
    }());
}

const sleep = t => new Promise(resolve => {
    setTimeout(resolve, t);
});

const start = Date.now();
const log = x => console.log(`${x} @${Date.now() - start}`);

async function* produceSlowly(n) {
    for (let i = 0; i < n; i++) {
        await sleep(2000);
        log(`i=${i}`);
        yield {nested: sleep(i * 1000).then(() => i)};
    }
}

(async function () {
    const i = limitConcurrent(produceSlowly(9), 2);
    for await (const x of i) {
        log(`         x=${x}`);
    }
})();
```

> Maybe it can be reduced to a simpler fix on the existing code?

I'm sure it can. I'll try to come up with something.

**Comment by bergus** [comment]
Here's what I'd do:
```ts
export function waitRaceAsync<T>(
    iterable: AsyncIterable<Promise<T> | T>,
    cacheSize: number
): AsyncIterable<T> {
    cacheSize = cacheSize > 1 ? cacheSize : 1; // cache size cannot be smaller than 1
    return {
        [$A](): AsyncIterator<T> {
            const source = iterable[$A]();
            let finished = false;
            // resolvers for currently active tasks,  which are racing to remove and call the first one
            const resolvers: null | ((res: IteratorResult<T> | Promise<never>) => void)[] = [];
            // cache of promises to be resolved or to be returned by `.next()` to the destination
            const promises: Promise<IteratorResult<T>>[] = [];
            
            function kickOffNext(): void {
                promises.push(new Promise((resolve, reject) => {
                    resolvers.push(resolve);
                    
                    // `new Promise` executor handles synchronous exceptions
                    source.next().then(a => {
                        if (a.done) {
                            finished = true;
                            resolvers.pop()(a);
                        } else if (isPromiseLike(a.value)) {
                            const promise = Promise.resolve(a.value);
                            promise.then(value => {
                                resolvers.shift()({done: false, value});
                                kickOffMore();
                            }, _ => {
                                resolvers.shift()(promise);
                                kickOffMore();
                            });
                        } else {
                            resolvers.shift()(a);
                        }
                        kickOffMore(); // advance source iterator as far as possible within limit
                    }, err => {
                        // handle rejections from calling `i.next()`
                        resolvers.shift()(Promise.reject(err));
                        finished = true; // ???
                    });
                });
            }
            function kickOffMore() {
                if (!finished // stop when source is done
                    && promises.length < cacheSize // backpressure: don't keep too many promises in the cache if destination doesn't poll `.next()` fast enough
                    && resolvers.length < cacheSize // limit: don't continue to more active tasks than the maximum
                ) {
                    kickOffNext();
                }
            }
            return {
                next(): Promise<IteratorResult<T>> {
                    if (!promises.length) {
                        kickOffNext();
                    }
                    return promises.shift();
                },
            };
        },
    };
}
```

There may be a problem if the returned iterator is polled too fast (not waiting for the promise returned by `.next()` before calling it again), currently this will just go beyond the limit but I think that's the best (and easiest) way to handle this - a well-behaved consumer should never do this anyway.

**Comment by vitaly-t** [comment]
@bergus Thank you!!! This does work, and passes the test that I created! 👍 

@RebeccaStevens I have the latest, using what @bergus gave us, inside `182-bugfix` branch, which works perfectly.

It just needs some rework for proper types control, and that's it!

**Comment by vitaly-t** [comment]
The code written by @bergus has been merged into the master, refactored somewhat.

Further refactoring is possible.

**Comment by bergus** [comment]
Oh well. I write

> a well-behaved consumer should never do this anyway.

but then actually that's exactly what the code does with `source`. If the promises contained in the `source` iterator resolve faster than the `source` iterator itself advanced, the `kickOffMore()` function will be called multiple times - and `source.next()` will be called again before the previous `source.next()` promise fulfilled.

This may require some more thought.

**Comment by vitaly-t** [comment]
@bergus Thank you for your contribution! If you can improve the code further, you are very welcome!

**Comment by vitaly-t** [comment]
In the meantime, the main issue has been addressed, so I am closing the issue. If something comes up again - we can open a new one ;)

Released in [v2.2.2](https://github.com/vitaly-t/iter-ops/releases/tag/2.2.2).

**Comment by vitaly-t** [comment]
Open related update ticket - #186

**Comment by vitaly-t** [comment]
@bergus I couldn't come up with a test that would make use of this piece of code:

```ts
                            (err) => {
                                // handle rejections from calling `i.next()`
                                resolvers.shift()!(Promise.reject(err));
                                finished = true;
                            }
```

Are you sure this section should ever execute? if so, how do you think it should be tested? I have added all the exception-throwing tests that I could think of, but none execute that code block.

**Comment by bergus** [comment]
Start with an actual async generator, instead of mapping a synchronous iterator to a promise :-)
`async function* oops() { yield 1; await sleep(10); throw new Error('boom'); }` will reject on the second `.next()` call.

**Comment by vitaly-t** [comment]
@bergus Thank you!!! [This commit](async function* oops() { yield 1; await sleep(10); throw new Error('boom'); }) did it 😄
